src/kernels/gpu_reference_kernel/naive_conv.cpp

```diff
@@ -1233,6 +1507,137 @@ inline __device__ void naive_conv_wrw_ndhwc(const src_data_t* __restrict__ p_in,
+                                                      Strides6D out_strides,
+                                                      int di,
+                                                      int hi,
+                                                      int wi,
+                                                      int n,
+                                                      int k_per_group,
+                                                      int c_per_group,
+                                                      int do_,
+                                                      int ho,
+                                                      int wo,
+                                                      int sz,
+                                                      int sy,
+                                                      int sx,
+                                                      int dz,
+                                                      int dy,
+                                                      int dx,
+                                                      int pz,
+                                                      int py,
+                                                      int px,
+                                                      int fz,
+                                                      int fy,
+                                                      int fx,
+                                                      int group)
+{
+    /*
+     *  need to compute total filter pixel: `group * k_per_group * fz * fy * fx
+     * * c_per_group`.
+     *  to distribute this workload, let one workgroup compute `fz * fy * fx *
+     * c_per_group` pixel,
+     *  hence need `group * k_per_group` workgroups (grid_size).
+     */
+    int k             = k_per_group * group;
+    int c             = c_per_group * group;
+    int thread_length = fz * fy * fx * c_per_group;
+    int bid           = blockIdx.x;
+    int ik            = bid % k_per_group;
+    int ig            = bid / k_per_group;
+
+    // p_in += static_cast<size_t>(ig) * c_per_group;
+    //
+    // p_wei += static_cast<size_t>(ig) * k_per_group * fz * fy * fx * c_per_group +
+    // static_cast<size_t>(ik) * fz * fy * fx * c_per_group;
+    //
+    // p_out += static_cast<size_t>(ig) * k_per_group + static_cast<size_t>(ik);
+
+    p_in += static_cast<size_t>(ig) * in_strides[1];
+
+    p_wei += static_cast<size_t>(ig) * wei_strides[5] + static_cast<size_t>(ik) * wei_strides[4];
+
+    p_out += static_cast<size_t>(ig) * out_strides[1] + static_cast<size_t>(ik) * out_strides[0];
+
+    for(int tid = threadIdx.x; tid < thread_length; tid += blockDim.x)
+    {
+        // TODO(Amber): these might need modification due to strides
+        int ic = tid % c_per_group;
+        int ix = (tid / c_per_group) % fx;
+        int iy = (tid / (c_per_group * fx)) % fy;
+        int iz = (tid / (c_per_group * fx * fy));
+
+        double value = .0f;
```

Comment:
It is suspicious that we've casted `src_data_t` to `acc_data_t` in the loop and we've got dedicated data to accumulator, but `value` is hardcoded `double`. There is no point in `acc_data_t` if it is silently casted everything to `double`.
Moreover later we cast `double` from `acc_data_t` to `dst_data_t`. It looks like the `value` must be `acc_data_t`.

```suggestion
        auto value = static_cast<acc_data_t>(0);
```

It applies to fwb and bwd as well.

Reply:
good catch. A lot of this stuff is from the original code. I'm planning a clean up after this PR just so that I can simplify debugging for myself in case there's a mismatch between packed and non-packed kernels.

Reply:
all instances have been fixed.